proxy/http2/Http2ConnectionState.cc

```diff
@@ -645,8 +645,21 @@ rcv_goaway_frame(Http2ConnectionState &cstate, const Http2Frame &frame)
   Http2StreamDebug(cstate.ua_session, stream_id, "GOAWAY: last stream id=%d, error code=%d", goaway.last_streamid,
                    static_cast<int>(goaway.error_code));
 
-  cstate.rx_error_code = {ProxyErrorClass::SSN, static_cast<uint32_t>(goaway.error_code)};
-  cstate.handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  // If peer indicated an error, go ahead and shutdown hard
+  if (goaway.error_code != Http2ErrorCode::HTTP2_ERROR_NO_ERROR) {
+    cstate.rx_error_code = {ProxyErrorClass::SSN, static_cast<uint32_t>(goaway.error_code)};
+    cstate.send_goaway_frame(cstate.get_latest_stream_id_in(), Http2ErrorCode::HTTP2_ERROR_NO_ERROR);
+    cstate.ua_session->set_half_close_local_flag(true);
+    cstate.handleEvent(HTTP2_SESSION_EVENT_FINI, nullptr);
+  } else {
+    // Otherwise, just make a note that we should not initiate any streams above goaway.last_streamid;
+    cstate.set_last_initiating_streamid(goaway.last_streamid);
+    // If last initiating stream_id is set to 0, then just disable push for the remainder of the session
+    if (goaway.last_streamid == 0) {
+      cstate.client_settings.set(HTTP2_SETTINGS_ENABLE_PUSH, 0);
```

Comment:
I understand why it is needed and it would work as expected, however, I'm not so keen to use `cstate.client_settings` here because these settings are values advertised on SETTINGS frame. It would be nice if we could use something else.